File: src/main/hardplayer/BasePlayer.scala
package hardplayer

import battlecode.common._
import hardplayer.Static._
import hardplayer.goal.Goal

abstract class BasePlayer(myRC : RobotController) extends Static {

	init(myRC)

	var goals : Array[Goal]

	def runloop() {
		sensorAI.sense()
		goals.foreach { g => g.execute() }
	}

	def repurpose() : Boolean = false

	def run() {
		while(true) {
			try {
				runloop()
				if(repurpose())
					return
			} catch {
				case e : Exception => { debug_stackTrace(e) }
			}
		}
		myRC.`yield`()
	}

}

File: src/main/mediumbot/BasePlayer.java
package mediumbot;

import battlecode.common.GameActionException;
import battlecode.common.MapLocation;
import battlecode.common.RobotController;
import battlecode.common.RobotType;
import battlecode.common.Team;

public abstract class BasePlayer {
	RobotController rc;
	RobotType myType;
	Team myTeam;
	Team neutralTeam;
	Team enemyTeam;
	int width;
	int height;
	MapLocation curLoc;
	double curHP;
	double curShields;
	
	public BasePlayer(RobotController rc) {
		this.rc = rc;
		this.myType = rc.getType();
		this.myTeam = rc.getTeam();
		this.neutralTeam = Team.NEUTRAL;
		this.enemyTeam = myTeam.opponent();
		width = rc.getMapWidth();
		height = rc.getMapHeight();
		updateCurVariables();
	}
	
	public abstract void run() throws GameActionException;
	
	public void loop() {
		while(true) {
			try {
				rc.setIndicatorString(0, "Delay: "+rc.roundsUntilActive());
				
				// Update variables
				updateCurVariables();
				
				// Execute turn
				run();
				
			} catch (Exception e) {
				e.printStackTrace();
			}
			
			// End turn
			rc.yield();
		}
	}
	
	void updateCurVariables() {
		curLoc = rc.getLocation();
		curHP = rc.getEnergon();
		curShields = rc.getShields();
	}
}
